### PR TITLE
fix build against nix 2.7

### PR DESCRIPTION
**Pull request by ajs124**
Includes  #1161 (@Ma27)

not particularly well tested, but builds and tests pass (I think)

### CODE REVIEW

File: src/hydra-eval-jobs/hydra-eval-jobs.cc [hunk 1]
```diff
@@ -157,7 +157,8 @@ static void worker(
                 if (drv->querySystem() == "unknown")
                     throw EvalError("derivation must have a 'system' attribute");
 
-                auto drvPath = drv->queryDrvPath();
+                auto localStore = state.store.dynamic_pointer_cast<LocalFSStore>();
+                auto drvPath = localStore->printStorePath(drv->requireDrvPath());
```
Comment (Ericson2314) [inline src/hydra-eval-jobs/hydra-eval-jobs.cc:161]: `localStore` might be `null`, please do not do this way.
```suggestion
                auto drvPath = state.store->printStorePath(drv->requireDrvPath());
```

File: src/hydra-eval-jobs/hydra-eval-jobs.cc [hunk 2]
```diff
@@ -201,7 +202,6 @@ static void worker(
                 /* Register the derivation as a GC root.  !!! This
                    registers roots for jobs that we may have already
                    done. */
-                auto localStore = state.store.dynamic_pointer_cast<LocalFSStore>();
```
Comment (Ericson2314) [inline src/hydra-eval-jobs/hydra-eval-jobs.cc:204]: Please put back

File: src/hydra-eval-jobs/hydra-eval-jobs.cc [hunk 3]
```diff
@@ -210,7 +210,7 @@ static void worker(
 
                 nlohmann::json out;
                 for (auto & j : outputs)
-                    out[j.first] = j.second;
+                    out[j.first] = localStore->printStorePath(j.second);
```
Comment (Ericson2314) [inline src/hydra-eval-jobs/hydra-eval-jobs.cc:213]:
```suggestion
                    out[j.first] = state.store->printStorePath(j.second);
```

**Review by Ma27 — APPROVED** [review]
Deployed to my own Hydra, seems to work fine :+1:

**Review by Ericson2314 — APPROVED** [review]

**Comment by Ericson2314** [comment]
@edolstra Feel comfortable merging this one?

**Comment by cole-h** [comment]
I feel like we should hold off on this, due to the recent issues with Hydra that maybe are connected to the bump to 2.6 (e.g. see the Infra and Infra alerts channels on Matrix, where this has been a ~hot topic).

**Comment by Ericson2314** [comment]
Oh, I didn't know about that. Bummer. Perhaps we need a way to disable non-blocking GC until the kinks are worked out? (Note, they could be Hydra side, e.g. a lack of temp roots, too.)

Per #1166 I would say we need a staging branch for every version of Nix newer than the on on `master`, in addition to Nix HEAD.

So we could revert 2,6 on master, but then revert the revert on `staging-nix-2.6`, and merge that and this into `staging-nix-2.7`, and then make `staging-nix-head` or so on top of that.

**Comment by edolstra** [comment]
@cole-h Do you have more info on those issues?

**Comment by grahamc** [comment]
@edolstra could you, @rbvermaa, and I do a call some time to go over them? They're a bit hard to track down and I think we might have an easier time sorting through them together.

**Comment by rbvermaa** [comment]
I would prefer it if we can allow Hydra to build against 2.4 and up. We had to revert back to 2.4, because we were getting builds that kept hanging, which significantly disrupted our builds.

**Comment by Ericson2314** [comment]
@rbvermaa The Nix internals change too much for that to be possible at this time. But we could keep the branches tied to older Nix around afterwords.

**Comment by rbvermaa** [comment]
It seems surprising to me to not support the versions of nix that are provided in NixOS stable channel.

**Comment by Ericson2314** [comment]
@rbvermaa Hydra has always been built against a single version of Nix. The daemon protocol is meant to be interoperable, so other Nix versions for the builders will work. (And I have unmerged PRs to test that interoperability.)

Hydra previously was pinned to random Nix versions. In recent months, it is pinned to a stable version of Nix. This is definitely an improvement.

**Comment by grahamc** [comment]
Applied in #1188.

**Pull request closed without merging**